### PR TITLE
opencv3,opencv4: disable some unnecessary vendoring on Darwin

### PR DESCRIPTION
**Pull request by delroth**
## Description of changes

See also: https://github.com/opencv/opencv/blob/4.x/CMakeLists.txt#L222 & https://github.com/opencv/opencv/blob/3.4/CMakeLists.txt#L220

Ref #254798 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by delroth** [comment]
@ofborg build opencv3 opencv4

**Comment by delroth** [comment]
Maintainers, I think I'm going to need your help here:

- opencv4 tests seem to randomly fail on x86_64-linux? It [passed](https://logs.ofborg.org/?key=nixos/nixpkgs.256444&attempt_id=dcd64869-2c54-49fd-9f59-a2e22e9ea94d) in one ofborg invocation, [didn't pass](https://logs.ofborg.org/?key=nixos/nixpkgs.256444&attempt_id=f2b58913-14e1-457f-b884-5e0ba483e853) in the other, even though there isn't even a diff caused by this PR on x86_64-linux.
- `opencv4-tests` fails on Darwin with this PR even though `opencv4` itself builds fine, and I don't have access to a Darwin host to debug this further.

In case this wasn't clear in the PR description, this is CVE-2023-4863 remediation related.

**Comment by delroth** [comment]
cc @gbtb who touched the tests last AFAICT and might be able to help too :)

**Comment by gbtb** [comment]
> cc @gbtb who touched the tests last AFAICT and might be able to help too :)

Hi. When I've added test support for opencv4 in nixpkgs, I've also noticed that x86-64_darwin tests [had not been working properly](https://github.com/NixOS/nixpkgs/pull/208351#issuecomment-1368386700) and chose to ignore them [for that platform](https://github.com/NixOS/nixpkgs/blob/652f79f16469f6006877cb3dd7a1e6148cdaf9a9/pkgs/development/libraries/opencv/4.x.nix#L494)
Considering flakiness of the tests for x86_64-linux platform, I also had to ignore some categories of tests that failed for various reasons. But then it was stable both locally and on CI. I haven't looked how the tests behave on Hydra since then. Maybe we can ignore offending tests or the whole test suite by removing it from test list [here](https://github.com/gbtb/nixpkgs/blob/66efeba6775c493bf264e8652481042d22ad35d1/pkgs/development/libraries/opencv/tests.nix)

File: pkgs/development/libraries/opencv/4.x.nix [hunk 1]
```diff
@@ -424,6 +424,17 @@ stdenv.mkDerivation {
   ] ++ lib.optionals stdenv.isDarwin [
     "-DWITH_OPENCL=OFF"
     "-DWITH_LAPACK=OFF"
+
+    # Disable unnecessary vendoring that's enabled by default only for Darwin.
+    # Note that the opencvFlag feature flags listed above still take
+    # precedence, so we can safely list everything here.
+    "-DBUILD_ZLIB=ON"
```
Comment (risicle) [inline pkgs/development/libraries/opencv/4.x.nix:431]: I'm confused. https://docs.opencv.org/4.x/db/d05/tutorial_config_reference.html, section

> Options naming conventions

Suggests `-DBUILD_XYZ=ON` _enables_ the "3rdparty library" (which is what opencv seem to call bundled libraries)

Comment (delroth) [reply pkgs/development/libraries/opencv/4.x.nix:431]: Thanks for catching that, I did indeed flip all the logic :) oops!

To my defense, I have 0 access to a Darwin host to test this :(

**Comment by delroth** [comment]
@ofborg build opencv3 opencv4

**Comment by risicle** [comment]
`nixpkgs-review` reveals no new failures, macos 10.15.

**Review by risicle — APPROVED** [review]

**Comment by delroth** [comment]
In absence of more reviews from Darwin maintainers, merging. I'm kind of unhappy with the state of this package, the broken tests at HEAD make it really hard to figure out whether new regressions are happening. But given that the alternative is marking this as insecure, I'll take the increased risk chance here - please don't hesitate to revert (but then, fix the vendoring before EOW).

**Pull request merged**